Python/crossinterp.c

```diff
@@ -436,6 +442,51 @@ _PyObject_GetXIData(PyThreadState *tstate,
     return 0;
 }
 
+int
+_PyObject_GetXIData(PyThreadState *tstate,
+                    PyObject *obj, _PyXIData_t *xidata)
+{
+    return _get_xidata(tstate, obj, _PyXIDATA_XIDATA_ONLY, xidata);
+}
+
+int
+_PyObject_GetXIDataWithFallback(PyThreadState *tstate,
+                                PyObject *obj, xidata_fallback_t fallback,
+                                _PyXIData_t *xidata)
+{
+    switch (fallback) {
+        case _PyXIDATA_XIDATA_ONLY:
+            return _get_xidata(tstate, obj, fallback, xidata);
+        case _PyXIDATA_FULL_FALLBACK:
+            if (_get_xidata(tstate, obj, fallback, xidata) == 0) {
+                return 0;
+            }
+            PyObject *exc = _PyErr_GetRaisedException(tstate);
+            if (PyFunction_Check(obj)) {
+                if (_PyFunction_GetXIData(tstate, obj, xidata) == 0) {
+                    Py_DECREF(exc);
+                    return 0;
+                }
+                _PyErr_Clear(tstate);
+            }
+            // We could try _PyMarshal_GetXIData() but we won't for now.
+            if (_PyPickle_GetXIData(tstate, obj, xidata) == 0) {
+                Py_DECREF(exc);
+                return 0;
+            }
+            // Raise the original exception.
+            _PyErr_SetRaisedException(tstate, exc);
+            return -1;
+        default:
+#ifdef Py_DEBUG
+            Py_UNREACHABLE();
```

Comment:
The `#ifdef` looks like a typo.

Reply:
It's there to make sure we abort on debug builds, but we have a fallback for non-debug builds.

Reply:
I'd prefer an explicit `Py_FatalError()` then.  I think I can see this usage only in `crossinterp.c` (here, L1673, L1700).

Reply:
My intention is to not crash under normal usage, which is what `Py_FatalError()` would do.

Reply:
ah, `Py_FatalError()` instead of `Py_UNREACHABLE()`?  I'm okay with that.  It really should be unreachable through.

Reply:
...and `Py_UNREACHABLE()` is just a wrapper around `Py_FatalError()` under `Py_DEBUG`.

Reply:
I strongly prefer that `Py_UNREACHABLE()` be used when we care about the C compiler's optimization on non-debug builds.  I won't be surprised even if it stops the abortion on debug builds in the future.  Also, the double `Py_DEBUG` checks seem unnatural to me.

Reply:
That's fair.  I'll switch it to explicitly `Py_FatalError()`.

Reply:
done

Reply:
Thanks.  This is off-topic, but would one of the coming PRs address https://github.com/python/cpython/pull/133107#discussion_r2075974786?

Reply:
Yeah, I'll be addressing that.

Reply:
fixed